hikari/impl/rest.py

```diff
@@ -133,6 +134,115 @@ def acquire(self) -> aiohttp.BaseConnector:
+        client_secret: str,
+        *,
+        scopes: typing.Sequence[typing.Union[applications.OAuth2Scope, str]] = (
+            applications.OAuth2Scope.APPLICATIONS_COMMANDS_UPDATE,
+            applications.OAuth2Scope.IDENTIFY,
+        ),
+    ) -> None:
+        self._client_id = snowflakes.Snowflake(client)
+        self._client_secret = client_secret
+        self._exception: typing.Optional[errors.ClientHTTPResponseError] = None
+        self._expire_at = 0.0
+        self._lock = asyncio.Lock()
+        self._scopes = tuple(scopes)
+        self._token: typing.Optional[str] = None
+
+    @property
+    def client_id(self) -> snowflakes.Snowflake:
+        """ID of the application this token strategy authenticates with.
+
+        Returns
+        -------
+        hikari.snowflakes.Snowflake
+            ID of the application this token strategy authenticates with.
+        """
+        return self._client_id
+
+    @property
+    def _is_expired(self) -> bool:
+        return time.monotonic() >= self._expire_at
+
+    @property
+    def scopes(self) -> typing.Sequence[typing.Union[applications.OAuth2Scope, str]]:
+        """Scopes this token strategy authenticates for.
+
+        Returns
+        -------
+        typing.Sequence[typing.Union[hikari.applications.OAuth2Scope, builtins.str]]
+            The scopes this token strategy authenticates for.
+        """
+        return self._scopes
+
+    async def acquire(self, client: rest_api.RESTClient) -> str:
+        if self._token and not self._is_expired:
+            return self._token
+
+        async with self._lock:
+            if self._token and not self._is_expired:
+                return self._token
+
+            if self._exception:
+                # If we don't copy the exception then python keeps adding onto the stack each time it's raised.
+                raise copy.copy(self._exception) from None
+
+            try:
+                response = await client.authorize_client_credentials_token(
+                    client=self._client_id, client_secret=self._client_secret, scopes=self._scopes
+                )
+
+            except errors.ClientHTTPResponseError as exc:
+                if not isinstance(exc, errors.RateLimitedError):
+                    # If we don't copy the exception then python keeps adding onto the stack each time it's raised.
+                    self._exception = copy.copy(exc)
+
+                raise
+
+            # Expires in is lowered a bit in-order to lower the chance of a dead token being used.
+            self._expire_at = time.monotonic() + math.floor(response.expires_in.total_seconds() * 0.99)
+            self._token = f"{response.token_type} {response.access_token}"
+            return self._token
+
+    async def close(self) -> None:
+        return None
+
+    def invalidate(self, token: typing.Optional[str]) -> None:
+        if token == self._token:
+            self._expire_at = 0.0
+            self._token = None
```

Comment:
The docstring for this function says that if `None` is passed, it will be invalidated regardless, which is not being done here